sycl/include/CL/sycl/detail/buffer_impl.hpp

```diff
@@ -99,23 +99,52 @@ template <typename AllocatorT> class buffer_impl {
     }
   }
 
+  template <typename Iterator> struct is_const_iterator {
+    typedef typename std::iterator_traits<Iterator>::pointer pointer;
+    static const bool value =
+        std::is_const<typename std::remove_pointer<pointer>::type>::value;
+  };
+
+  template <typename Iterator>
+  using EnableIfConstIterator =
+      typename std::enable_if<is_const_iterator<Iterator>::value,
+                              Iterator>::type;
+
+  template <typename Iterator>
+  using EnableIfNotConstIterator =
+      typename std::enable_if<!is_const_iterator<Iterator>::value,
+                              Iterator>::type;
+
   template <class InputIterator>
-  buffer_impl(InputIterator first, InputIterator last, const size_t sizeInBytes,
-              const property_list &propList,
+  buffer_impl(EnableIfNotConstIterator<InputIterator> first, InputIterator last,
```

Comment:
Hm, it will not work because I will get two function templates that differ only in their default template arguments. (c'tors with EnableIf**Not**ConstIterator and EnableIfConstIterator). You can see note from here: https://en.cppreference.com/w/cpp/types/enable_if

Reply:
Ah, interesting, you found the exact explanation!
By looking at our code base, we managed the same use case in a similar way as you https://github.com/triSYCL/triSYCL/blob/master/include/CL/sycl/buffer/detail/buffer.hpp#L361
where we put the SFINAE in a spare argument.
But it also makes the function globally more complex by moving apart the SFINAE mechanics....
So I am fine with your solution too.